### ISSUE TITLE
Possible bug for +=

### ISSUE THREAD

**Issue by ranocha** [chriselrod/LoopVectorization.jl#194]
I stumbled across the following erroneous behavior when trying to speed up some code using LoopVectorization.
```julia
julia> using Pkg; Pkg.activate(temp=true); Pkg.add("LoopVectorization")
[...]
  [bdcacae8] LoopVectorization v0.11.0
[...]

julia> versioninfo()
Julia Version 1.5.3
Commit 788b2c77c1 (2020-11-09 13:37 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i7-8700K CPU @ 3.70GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, skylake)
Environment:
  JULIA_MPI_BINARY = system
  JULIA_NUM_THREADS = 6

julia> using LoopVectorization

julia> function foo!(du, u, mat)
           @assert size(u, 1) == size(u, 2) == size(mat, 1) == size(mat, 2)
           for i2 in 1:size(u, 2), i1 in 1:size(u, 1)
               for sum_idx in 1:size(u, 1)
                   du[i1, i2] += mat[i1, sum_idx] * u[sum_idx, i2] + mat[i2, sum_idx] * u[i1, sum_idx]
               end
           end
           return nothing
       end

julia> function foo_avx!(du, u, mat)
           @assert size(u, 1) == size(u, 2) == size(mat, 1) == size(mat, 2)
           @avx for i2 in 1:size(u, 2), i1 in 1:size(u, 1)
               for sum_idx in 1:size(u, 1)
                   du[i1, i2] += mat[i1, sum_idx] * u[sum_idx, i2] + mat[i2, sum_idx] * u[i1, sum_idx]
               end
           end
           return nothing
       end

julia> u = [1.0 2.0; 3.0 4.0]; du = zero(u); du_avx = zero(u); mat = [0.1 0.2; 0.3 0.4];

julia> foo!(du, u, mat); foo_avx!(du_avx, u, mat); du_avx ≈ du
false

julia> du
2×2 Array{Float64,2}:
 1.2  2.1
 2.6  4.7

julia> du_avx
2×2 Array{Float64,2}:
 1.0  1.6
 2.0  3.2
```
I get the same results on Julia v1.6.0-beta1 and v1.7.0-DEV.358.

Using `@avx` seems to be okay when I expand `+=` manually as follows.
```julia
julia> function foo_new!(du, u, mat)
           @assert size(u, 1) == size(u, 2) == size(mat, 1) == size(mat, 2)
           for i2 in 1:size(u, 2), i1 in 1:size(u, 1)
               dui1i2 = du[i1, i2]
               for sum_idx in 1:size(u, 1)
                   dui1i2 += mat[i1, sum_idx] * u[sum_idx, i2] + mat[i2, sum_idx] * u[i1, sum_idx]
               end
               du[i1, i2] = dui1i2
           end
           return nothing
       end

julia> function foo_avx_new!(du, u, mat)
           @assert size(u, 1) == size(u, 2) == size(mat, 1) == size(mat, 2)
           @avx for i2 in 1:size(u, 2), i1 in 1:size(u, 1)
               dui1i2 = du[i1, i2]
               for sum_idx in 1:size(u, 1)
                   dui1i2 += mat[i1, sum_idx] * u[sum_idx, i2] + mat[i2, sum_idx] * u[i1, sum_idx]
               end
               du[i1, i2] = dui1i2
           end
           return nothing
       end

julia> u = [1.0 2.0; 3.0 4.0]; du = zero(u); du_avx = zero(u); mat = [0.1 0.2; 0.3 0.4];

julia> foo_new!(du, u, mat); foo_avx_new!(du_avx, u, mat); du_avx ≈ du
true

julia> du
2×2 Matrix{Float64}:
 1.2  2.1
 2.6  4.7

julia> du_avx
2×2 Matrix{Float64}:
 1.2  2.1
 2.6  4.7
```
This is the same result as obtained by `foo!` but not by `foo_avx!` above.

Is this a bug in LoopVectorization or am I doing something stupid here?

**Comment by chriselrod** [comment]
Thanks for the issue!
Fixed by fe94ecfb259a801bb6153bc8e846e1f6b1c3fdcf and [added to the tests](https://github.com/chriselrod/LoopVectorization.jl/blob/b5c0af9ab00061cf2d650af04a97becc9b5ad495/test/gemm.jl#L576).

> Is this a bug in LoopVectorization or am I doing something stupid here?

Yes, it was a bug. For the most part, barring some exceptions like around around aliasing (where `@avx` makes less aggressive assumptions than `@simd ivdep`), if it works without `@avx` it should work with it.

LoopVectorization's internal representation of loops doesn't look very much like loops, and there can be/are bugs in its code to build the internal representation of relationships between the operations. This was one (now fixed). 
Basically, it was search recursively for dependence between variables, but not memory accesses. If you have
```julia
d = muladd(a, b, c)
c = muladd(e, f, d)
```
It would look through `d` to find you were doing `c = f(c, other_args)`.
But if you had
```julia
d = muladd(a, b, c[i])
c[i] = muladd(e, f, d)
```
It wasn't checking for `c[i]`.
Bugs like this come up because it doesn't reduce the expression to SSA form before converting to its internal structure.

One of the reasons I don't is because I'm concerned about compile times, and
```julia
julia> using LoopVectorization
q = :
julia> q = :(for i2 in 1:size(u, 2), i1 in 1:size(u, 1)
                      for sum_idx in 1:size(u, 1)
                          du[i1, i2] += mat[i1, sum_idx] * u[sum_idx, i2] + mat[i2, sum_idx] * u[i1, sum_idx]
                      end
                  end);

julia> @benchmark LoopVectorization.LoopSet(copy(q)) # convert `q` to LoopVectorization's internal representation
BenchmarkTools.Trial:
  memory estimate:  57.86 KiB
  allocs estimate:  744
  --------------
  minimum time:     21.714 μs (0.00% GC)
  median time:      23.188 μs (0.00% GC)
  mean time:        28.458 μs (14.78% GC)
  maximum time:     3.858 ms (97.24% GC)
  --------------
  samples:          10000
  evals/sample:     1

julia> @benchmark Meta.lower(LoopVectorization, q) # lower to SSA
BenchmarkTools.Trial:
  memory estimate:  14.28 KiB
  allocs estimate:  200
  --------------
  minimum time:     698.237 μs (0.00% GC)
  median time:      737.116 μs (0.00% GC)
  mean time:        774.344 μs (0.14% GC)
  maximum time:     4.234 ms (81.76% GC)
  --------------
  samples:          6449
  evals/sample:     1
```

**Comment by ranocha** [comment]
Thanks a lot!
